kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/config/Configuration.java

```diff
@@ -167,20 +167,24 @@ public Configuration(
         this(adminHttp, filterDefinitions, defaultFilters, virtualClusters, null, micrometer, useIoUring, development);
     }
 
-    public static VirtualClusterModel toVirtualClusterModel(@NonNull VirtualCluster virtualCluster,
-                                                            @NonNull PluginFactoryRegistry pfr,
-                                                            @NonNull List<NamedFilterDefinition> filterDefinitions,
-                                                            @NonNull String virtualClusterNodeName) {
+    private static VirtualClusterModel toVirtualClusterModel(@NonNull VirtualCluster virtualCluster,
+                                                             @NonNull PluginFactoryRegistry pfr,
+                                                             @NonNull List<NamedFilterDefinition> filterDefinitions,
+                                                             @NonNull String virtualClusterNodeName) {
 
         VirtualClusterModel virtualClusterModel = new VirtualClusterModel(virtualClusterNodeName,
                 virtualCluster.targetCluster(),
-                buildAddressProviderService(virtualCluster, pfr),
-                virtualCluster.tls(),
                 virtualCluster.logNetwork(),
                 virtualCluster.logFrames(),
                 filterDefinitions);
-        logVirtualClusterSummary(virtualClusterModel.getClusterName(), virtualClusterModel.targetCluster(), virtualClusterModel.getClusterNetworkAddressConfigProvider(),
-                virtualCluster.tls());
+
+        virtualCluster.listeners().forEach(listener -> {
+            var networkAddress = buildNetworkAddressProviderService(listener.clusterNetworkAddressConfigProvider(), pfr);
+            var tls = listener.tls();
+            virtualClusterModel.addListener(listener.name(), networkAddress, tls);
+            logVirtualClusterSummary(virtualClusterModel.getClusterName(), virtualClusterModel.targetCluster(), networkAddress,
```

Comment:
I haven't got far enough through the PR to check if the impl has updated, but calling `logVirtualClusterSummary` multiple times per virtual cluster sounds a bit wrong.

Reply:
Yeah, that sounds fishy.. I'll look.